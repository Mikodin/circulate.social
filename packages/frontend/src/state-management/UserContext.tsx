import { createContext } from 'react';
import { CognitoUser } from '@aws-amplify/auth';
import type { ConfirmSignUp } from '../types/amplify.d';

export type UserContextType = {
  user?: {
    email: string;
    firstName: string;
    lastName: string;
    timezone: string;
    id: string;
  };
  userIsLoaded?: boolean;
  jwtToken?: string;
  register: (username: string, password: string) => Promise<CognitoUser>;

  restoreUser: (bypassCache: boolean) => Promise<UserContextType['user']>;

  confirmEmail: (
    username: string,
    code: string,
    routeOnSuccess?: string
  ) => Promise<ConfirmSignUp>;

  resendRegisterCode: (username: string) => Promise<boolean>;

  signIn: (
    username: string,
    password: string
  ) => Promise<UserContextType['user']>;

  signOut: () => Promise<boolean>;

  forgotPasswordInit: (username: string) => Promise<boolean>;

  forgotPasswordSubmit: (
    username: string,
    code: string,
    newPassword: string
  ) => Promise<boolean>;

  getIsUserLoggedIn: () => boolean;

  updateUserAttributes: (
    firstName: string,
    lastName: string,
    timezone: string
  ) => Promise<string>;
};

const UserContext = createContext<UserContextType | undefined>(undefined);

export default UserContext;
