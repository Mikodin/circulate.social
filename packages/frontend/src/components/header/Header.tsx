import { useContext, Fragment } from 'react';
import Link from 'next/link';
import UserContext from '../../state-management/UserContext';

import css from './Header.module.scss';

const Header = (): JSX.Element => {
  const { signOut, getIsUserLoggedIn } = useContext(UserContext);
  return (
    <Fragment>
      <div className={css.container}>
        <div className={css.logoContainer}>
          <h2>
            <Link href="/">
              <a>Circulate.social</a>
            </Link>
          </h2>
        </div>
        <div className={css.actionsContainer}>
          {getIsUserLoggedIn() ? (
            <Fragment>
              <a onClick={signOut}>Logout</a>
              <Link href="/circles/home">
                <a>My Circles</a>
              </Link>
            </Fragment>
          ) : (
            <Fragment>
              <Link href="/start-a-circle">
                <a>Start a circle</a>
              </Link>
            </Fragment>
          )}
        </div>
      </div>
    </Fragment>
  );
};

export default Header;
