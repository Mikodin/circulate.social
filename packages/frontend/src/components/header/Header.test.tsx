import React from 'react';
import { render } from '@testing-library/react';
import UserContext from '../../state-management/UserContext';

import Header from './Header';

describe('It should pass', () => {
  it('Passes', () => {
    expect(true).toBe(true);
  });
});

function renderHeader(props?: any) {
  const defaultProps = {
    signOut: jest.fn(),
    getIsUserLoggedIn: jest.fn(() => false),
  };
  return render(
    <UserContext.Provider value={{ ...defaultProps, ...props }}>
      <Header />
    </UserContext.Provider>
  );
}

describe('Header', () => {
  describe('Regardless of user status', () => {
    it('Should show our logo', () => {
      const { getByText } = renderHeader();

      expect(getByText('Circulate.social')).toMatchInlineSnapshot(`
      <a
        href="/"
      >
        Circulate.social
      </a>
    `);
    });
  });
  describe('When user is signed out', () => {
    const renderSignedOutHeader = () =>
      renderHeader({ getIsUserLoggedIn: jest.fn(() => false) });
    it('Should match the snapshot', () => {
      const { container } = renderSignedOutHeader();
      expect(container).toMatchSnapshot();
    });

    it('Should display "Start a circle"', () => {
      const { queryByText } = renderSignedOutHeader();

      expect(queryByText('Start a circle')).toBeTruthy();
    });

    it('Should not show Logout button', () => {
      const { queryByText } = renderSignedOutHeader();

      expect(queryByText('Logout')).toBeFalsy();
    });

    it('Should not show My Circles button', () => {
      const { queryByText } = renderSignedOutHeader();

      expect(queryByText('My Circles')).toBeFalsy();
    });
  });
  describe('When user is signed in', () => {
    const renderSignedInHeader = () =>
      renderHeader({ getIsUserLoggedIn: jest.fn(() => true) });
    it('Should match the snapshot', () => {
      const { container } = renderSignedInHeader();
      expect(container).toMatchSnapshot();
    });

    it('Should show Logout button', () => {
      const { queryByText } = renderSignedInHeader();

      expect(queryByText('Logout')).toBeTruthy();
    });

    it('Should show My Circles button', () => {
      const { queryByText } = renderSignedInHeader();

      expect(queryByText('My Circles')).toBeTruthy();
    });

    it('Should not display "Start a circle"', () => {
      const { queryByText } = renderSignedInHeader();

      expect(queryByText('Start a circle')).toBeFalsy();
    });
  });
});
