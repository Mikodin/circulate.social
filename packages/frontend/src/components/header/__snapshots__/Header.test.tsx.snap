// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header When user is signed in Should match the snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="logoContainer"
    >
      <h2>
        <a
          href="/"
        >
          Circulate.social
        </a>
      </h2>
    </div>
    <div
      class="actionsContainer"
    >
      <a>
        Logout
      </a>
      <a
        href="/circles/home"
      >
        My Circles
      </a>
    </div>
  </div>
</div>
`;

exports[`Header When user is signed out Should match the snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="logoContainer"
    >
      <h2>
        <a
          href="/"
        >
          Circulate.social
        </a>
      </h2>
    </div>
    <div
      class="actionsContainer"
    >
      <a
        href="/start-a-circle"
      >
        Start a circle
      </a>
    </div>
  </div>
</div>
`;
